### PR TITLE
Better support for HTMLMediaElement in the loadPromise helper.

### PR DESCRIPTION
**Pull request by gmajoulet**
This PR ensures that `amp-video` and `amp-audio` `layoutCallback`s complete and don't remain pending.

- Changing `loadstart` to `loadedmetadata` to resolve the `loadPromise`

`amp-video` and `amp-audio` rely on `loadPromise` to complete their `layoutCallback`. This PR improves the `isLoaded` helper so it correctly detects that the HTMLMediaElement already triggered the event `loadPromise` would be waiting for.
Without this, `amp-video` and `amp-audio` `layoutCallback` could potentially never resolve in specific contexts, like when handled by the amp-story media pool.

The code used to listen to the `loadstart` event, but there is no way to synchronously check if this event already fired. We switch to listening to the `loadedmetadata` event that can be checked synchronously with `readyState > 0`.

- Handling `source` children

From [the spec](https://html.spec.whatwg.org/multipage/embedded-content.html#the-source-element): _"If a source element is inserted as a child of a media element that has no src attribute [...] the user agent must invoke the media element's resource selection algorithm."_.
If the `HTMLMediaElement` has no `src`, we listen to `loadedmetadata` and `error` events on the `source` elements.
If a `source` provides an unsupported format, or fails to load, it will trigger an error. But the following `source` element could load. So we listen for the `error` event on the last `source` element, but for `loadedmetadata` on any.

- Handling `preload="none"`

If the `HTMLMediaElement` specified `preload="none"`, we resolve the `layoutCallback` right away, without waiting for the `loadPromise` that wouldn't resolve before the user interacts with the media.

I believe this should live in `layoutCallback` as it could be misleading for `loadPromise` to resolve before the media actually loaded.

- Using a local URL for the `amp-story` tests, copying the error handling used in `test-amp-video.js`

Fixes https://github.com/ampproject/amphtml/issues/21794

### CODE REVIEW

**Comment by gmajoulet** [comment]
Of course!

I hosted these:
[Story demo link](https://video-loadedmetadata.firebaseapp.com/examples/s20/body-painting/index.html) (page 2 and 4 have animate-in videos)
[amp-video examples link](https://video-loadedmetadata.firebaseapp.com/examples/amp-video.amp.html)

File: src/event-helper.js [hunk 1]
```diff
@@ -131,6 +131,7 @@ export function listenOncePromise(element, eventType, opt_evtListenerOpts,
  */
 export function isLoaded(eleOrWindow) {
   return !!(eleOrWindow.complete || eleOrWindow.readyState == 'complete'
+      || (eleOrWindow instanceof HTMLMediaElement && eleOrWindow.readyState > 0)
```
Comment (jridgewell) [inline src/event-helper.js]: `instanceof` will break across FIE boundaries. Use `typeof eleOrWindow.readyState === 'number'` instead.

File: src/event-helper.js [hunk 2]
```diff
@@ -154,14 +155,13 @@ export function loadPromise(eleOrWindow) {
   const loadingPromise = new Promise((resolve, reject) => {
     // Listen once since IE 5/6/7 fire the onload event continuously for
     // animated GIFs.
-    const {tagName} = eleOrWindow;
-    if (tagName === 'AUDIO' || tagName === 'VIDEO') {
-      unlistenLoad = listenOnce(eleOrWindow, 'loadstart', resolve);
+    if (eleOrWindow instanceof HTMLMediaElement) {
```
Comment (jridgewell) [inline src/event-helper.js]: What elements are `HTMLMediaElement`?

Comment (gmajoulet) [reply src/event-helper.js]: `audio` and `video`, I thought this would be a no-op. I can revert, and use this same test in `isLoaded`

File: src/event-helper.js (hunk 1 above)
Comment (aghassemi) [reply src/event-helper.js]: OMG, catch of the year. Took me 5 mins to even understand why. 👏 🙇 @jridgewell

Comment (gmajoulet) [reply src/event-helper.js]: Thanks!! Done

File: src/event-helper.js (hunk 2 above)
Comment (gmajoulet) [reply src/event-helper.js]: Done

File: src/event-helper.js (hunk 1 above)
Comment (jridgewell) [reply src/event-helper.js]: Oh, I probably should have explained. FIEs are different HTML documents, and have different `window`s, and that window has a difference `HTMLMediaElement` function instance. So `fie.querySelector('audio') instanceof fie.parent.HTMLMediaElement` is always false.

**Comment by gmajoulet** [comment]
_(copy/pasting this to the PR description too)_

Additional changes to ensure that `layoutCallback` completes and doesn't remain pending:

- Handling `source` children

From [the spec](https://html.spec.whatwg.org/multipage/embedded-content.html#the-source-element): _"If a source element is inserted as a child of a media element that has no src attribute [...] the user agent must invoke the media element's resource selection algorithm."_.
If the `HTMLMediaElement` has no `src`, we listen to `loadedmetadata` and `error` events on the `source` elements.
If a `source` provides an unsupported format, or fails to load, it will trigger an error. But the following `source` element could load. So we listen for the `error` event on the last `source` element, but for `loadedmetadata` on any.

- Handling `preload="none"`

If the `HTMLMediaElement` specified `preload="none"`, we resolve the `layoutCallback` right away, without waiting for the `loadPromise` that wouldn't resolve before the user interacts with the media.

- Using a local URL for the `amp-story` tests, copying the error handling used in `test-amp-video.js`

PTAL

File: src/event-helper.js [hunk 3]
```diff
@@ -152,18 +153,32 @@ export function loadPromise(eleOrWindow) {
...
   const loadingPromise = new Promise((resolve, reject) => {
+    const isMediaElement = isHTMLMediaElement(eleOrWindow);
     // Listen once since IE 5/6/7 fire the onload event continuously for
     // animated GIFs.
-    const {tagName} = eleOrWindow;
-    if (tagName === 'AUDIO' || tagName === 'VIDEO') {
-      unlistenLoad = listenOnce(eleOrWindow, 'loadstart', resolve);
+    if (isMediaElement) {
+      // The following event can be triggered by the media or one of its
+      // sources. Using capture is required as the media events do not bubble.
```
Comment (cvializ) [inline src/event-helper.js]: 📝I was looking for a source for this and it looks like `loadedmetadata` ([1]) and `error` ([2]) and some others are *simple events* which are defined to not bubble ([3])

[1]: https://www.w3.org/TR/2011/WD-html5-20110113/video.html#fire-loadedmetadata
[2]: https://www.w3.org/TR/2011/WD-html5-20110113/video.html#fatal-decode-error
[3]: https://www.w3.org/TR/2011/WD-html5-20110113/webappapis.html#fire-a-simple-event

File: src/event-helper.js [hunk 4]
```diff
@@ -152,18 +153,32 @@ export function loadPromise(eleOrWindow) {
...
+      unlistenLoad =
+          listenOnce(eleOrWindow, 'loadedmetadata', resolve, {capture: true});
```
Comment (jridgewell) [inline src/event-helper.js]: Why capture?

File: src/event-helper.js [hunk 5]
```diff
@@ -152,18 +153,32 @@ export function loadPromise(eleOrWindow) {
...
+    // Don't unlisten on error for Windows.
+    if (!eleOrWindow.tagName) {
+      return;
     }
+    let errorTarget = eleOrWindow;
+    // If the media element has no `src`, it will try to load the sources in
+    // document order. If the last source errors, then the media element
+    // loading errored.
+    if (isMediaElement && !eleOrWindow.hasAttribute('src')) {
+      errorTarget = eleOrWindow.querySelector('source:last-of-type');
```
Comment (jridgewell) [inline src/event-helper.js]: Can source elements be nested inside another element?

File: src/event-helper.js [hunk 6]
```diff
@@ -152,18 +153,32 @@ export function loadPromise(eleOrWindow) {
...
+      if (!errorTarget) {
+        return reject('Media has no source.');
```
Comment (jridgewell) [inline src/event-helper.js]: Rejections must be Error instances.

File: extensions/amp-audio/0.1/amp-audio.js [hunk 7]
```diff
@@ -129,6 +129,11 @@ export class AmpAudio extends AMP.BaseElement {
       artwork: [{src: artwork}],
     };
 
+    // Resolve layoutCallback right away if the audio won't preload.
+    if (this.element.getAttribute('preload') === 'none') {
```
Comment (jridgewell) [inline extensions/amp-audio/0.1/amp-audio.js]: I'm not sure I understand this change. Can you explain what the issue is?

Comment (gmajoulet) [reply extensions/amp-audio/0.1/amp-audio.js]: The `preload` attribute set to `none` will delay the loading of the video/audio until user clicks play. The `loadPromise` would be pending until then, preventing the `layoutCallback` from resolving.
TLDR: this change ensures the `layoutCallback` doesn't wait for a user to click "play" to resolve.

File: src/event-helper.js (hunk 5 above)
Comment (jridgewell) [reply src/event-helper.js]: I meant for `<audio>` and `<video>` specifically:

```html
<video width="320" height="240" controls>
  <div>
    <source src="movie.mp4" type="video/mp4">
  </div>
  <div>
    <source src="movie.ogg" type="video/ogg">
  </div>
  Your browser does not support the video tag.
</video>
```

It seems it has to be a direct child for it work, so we can switch to https://github.com/ampproject/amphtml/blob/9ae3a653ba00dc9a3099bfdd60ae550230ecf7b4/src/dom.js#L367-L381 to get a speed boost.

Comment (gmajoulet) [reply src/event-helper.js]: Cool! Done
